test/functionapp.test.ts

```diff
@@ -0,0 +1,125 @@
+                // If the test failed, the resource group might not actually exist
+                console.log(`Ignoring resource group "${resourceGroup}" because it does not exist.`);
+            }
+        }
+        ext.tree.dispose();
+    });
+
+    test('createFunctionApp (Basic)', async () => {
+        const resourceName2: string = getRandomHexString().toLowerCase();
+        resourceGroupsToDelete.push(resourceName2);
+        await vscode.workspace.getConfiguration(extensionPrefix).update('advancedCreation', false, vscode.ConfigurationTarget.Global);
+        await updateGlobalSetting(projectLanguageSetting, ProjectLanguage.JavaScript);
+        ext.ui = new TestUserInput([resourceName2]);
+        await vscode.commands.executeCommand('azureFunctions.createFunctionApp');
+        const client: WebSiteManagementClient = getWebsiteManagementClient(testAccount);
+        const createdApp: WebSiteManagementModels.Site = await client.webApps.get(resourceName2, resourceName2);
+        assert.ok(createdApp);
+    });
+
+    test('createFunctionApp (Advanced)', async () => {
+        resourceGroupsToDelete.push(resourceName1);
+        await vscode.workspace.getConfiguration(extensionPrefix).update('advancedCreation', true, vscode.ConfigurationTarget.Global);
+        const testInputs: string[] = [resourceName1, 'Linux', '.NET', '$(plus) Create new resource group', resourceName1, '$(plus) Create new storage account', resourceName1, 'West US'];
+        ext.ui = new TestUserInput(testInputs);
+        await vscode.commands.executeCommand('azureFunctions.createFunctionApp');
+        const client: WebSiteManagementClient = getWebsiteManagementClient(testAccount);
+        const createdApp: WebSiteManagementModels.Site = await client.webApps.get(resourceName1, resourceName1);
+        assert.ok(createdApp);
+    });
+
+    test('stopFunctionApp', async () => {
+        ext.ui = new TestUserInput([resourceName1]);
+        await vscode.commands.executeCommand('azureFunctions.stopFunctionApp');
+        const client: WebSiteManagementClient = getWebsiteManagementClient(testAccount);
+        const createdApp: WebSiteManagementModels.Site = await client.webApps.get(resourceName1, resourceName1);
+        assert.equal(createdApp.state, 'Stopped', `Function App state should be 'Stopped' rather than ${createdApp.state}.`);
+    });
+
+    test('startFunctionApp', async () => {
+        ext.ui = new TestUserInput([resourceName1]);
+        await vscode.commands.executeCommand('azureFunctions.startFunctionApp');
+        const client: WebSiteManagementClient = getWebsiteManagementClient(testAccount);
+        const createdApp: WebSiteManagementModels.Site = await client.webApps.get(resourceName1, resourceName1);
+        assert.equal(createdApp.state, 'Running', `Function App state should be 'Running' rather than ${createdApp.state}.`);
+    });
+
+    test('restartFunctionApp', async () => {
+        let client: WebSiteManagementClient;
```

Comment:
Usually a restart occurs from a running state, not a stopped state. I suggest removing this logic to stop the function app and instead verifying that the app is "Running" before and after you call `'azureFunctions.restartFunctionApp'`

Reply:
I am confused for this comment, if verifing the app is "Running" before and after we call 'azureFunctions.restartFunctionApp', It cannot  exclud this situation "The 'azureFunctions.restartFunctionApp' execution fails, but the result is still pass".

Reply:
I agree it's not perfect, but my suggestion is simpler and covers the more common case. Your existing test is essentially testing "start" instead of "restart".

If you want the test to be completely accurate, you would poll for the status while `azureFunctions.restartFunctionApp` occurs to make sure it goes to 'Stopped' state in the middle. You're welcome to try that, but I think that code will be pretty complex and isn't worth it for now.